### PR TITLE
[libc] Add `__builtin_expect` tag on assert conditions; NFC

### PR DESCRIPTION
**Pull request by goldsteinn**

### CODE REVIEW

**Review by jhuber6 — COMMENTED** [review]
I remember discussing something similar with @AaronBallman about adding `__builtin_assume` to the assert macros. I think the answer then was that the C standard specifies exactly what `assert` should look like so we can't change it? I think `LIBC_ASSERT` probably should though.

**Comment by goldsteinn** [comment]
> I remember discussing something similar with @AaronBallman about adding `__builtin_assume` to the assert macros. I think the answer then was that the C standard specifies exactly what `assert` should look like so we can't change it? I think `LIBC_ASSERT` probably should though.

`__builtin_assume` is a bit different, no? This is properly `NFC`, it just has an off chance of helping codegen.
I don't think any standard says much about how `assert` is implemented other than it will `abort` and maybe write something to `stderr` (and I guess the linux reference saying `__assert_fail` is defined) but this doesn't change any of those things.

**Comment by jhuber6** [comment]
> > I remember discussing something similar with @AaronBallman about adding `__builtin_assume` to the assert macros. I think the answer then was that the C standard specifies exactly what `assert` should look like so we can't change it? I think `LIBC_ASSERT` probably should though.
> 
> `__builtin_assume` is a bit different, no? This is properly `NFC`, it just has an off chance of helping codegen. I don't think any standard says much about how `assert` is implemented other than it will `abort` and maybe write something to `stderr` (and I guess the linux reference saying `__assert_fail` is defined) but this doesn't change any of those things.

> Nope -- you cannot expand the argument given to assert if NDEBUG is defined (also, 7.2p1 defines what assert is required to expand to in this case.)

Was the response I got earlier, but I guess that only applies to making `#define assert(cond) __builtin_assume(!cond);` so this might be fine.

File: libc/src/__support/libc_assert.h
Comment (lntue) [inline libc/src/__support/libc_assert.h:74]: Can you also include `src/__support/macros/optimization.h` where these macros are defined and update cmake dependency?  Same for `libc/src/assert/assert.h`

**Comment by AaronBallman** [comment]
> > > I remember discussing something similar with @AaronBallman about adding `__builtin_assume` to the assert macros. I think the answer then was that the C standard specifies exactly what `assert` should look like so we can't change it? I think `LIBC_ASSERT` probably should though.
> > 
> > 
> > `__builtin_assume` is a bit different, no? This is properly `NFC`, it just has an off chance of helping codegen. I don't think any standard says much about how `assert` is implemented other than it will `abort` and maybe write something to `stderr` (and I guess the linux reference saying `__assert_fail` is defined) but this doesn't change any of those things.
> 
> > Nope -- you cannot expand the argument given to assert if NDEBUG is defined (also, 7.2p1 defines what assert is required to expand to in this case.)
> 
> Was the response I got earlier, but I guess that only applies to making `#define assert(cond) __builtin_assume(!cond);` so this might be fine.

Yeah, I think this is okay. One thing I did note when I reviewed the code is that the definition of `assert` changed in C23, it's now defined as `assert(...)` rather than `assert(e)`, see https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2829.htm for details.

**Comment by nickdesaulniers** [comment]
Whats the status of this PR? Looks like it's good to go?

**Comment by goldsteinn** [comment]
> Whats the status of this PR? Looks like it's good to go?

Still waiting on approval.

**Review by lntue — APPROVED** [review]

**Pull request merged**
